src/index.js

```diff
@@ -0,0 +1,324 @@
+  }
+
+  /**
+   * If the "master" lock is acquired, this worker will be responsible for
+   * assigning oplogs to workers
+   */
+  acquireMasterLock() {
+    const key = `${this.keyPrefix}:master`;
+    return this.lock.acquire(key, {
+      timeout: this.masterDuration * 1000,
+      retries: 0
+    })
+      .then(() => {
+        debug(`${this.workerId} is master.`);
+        return this.assignWorkers();
+      })
+      // ignore lock acquisition errors
+      .catch(err => {});
+  }
+
+  registerWorker() {
+    const key = this.workerRegistrationKey;
+    const now = Date.now();
+    return this.redisClient.zaddAsync(key, now, this.workerId)
+      .then(() => this.startNewAssignments())
+  }
+
+  getWorkerIds() {
+    const expiredTime = Date.now() - (this.healthcheckDuration * 1000);
+    const key = this.workerRegistrationKey;
+    return Promise.resolve()
+      // delete workers that have failed to renew their registration
+      .then(() => this.redisClient.zremrangebyscoreAsync(key, '-inf', expiredTime))
+      // get active workers
+      .then(() => this.redisClient.zrangeAsync(key, 0, -1));
+  }
+
+  getWorkerAssignmentsKey(workerId) {
+    return `${this.keyPrefix}:worker:${workerId}`;
+  }
+
+  readWorkerAssignmentsFromRedis(workerId) {
+    const key = this.getWorkerAssignmentsKey(workerId);
+    return this.redisClient.smembersAsync(key);
+  }
+
+  writeWorkerAssignmentsToRedis() {
+    return Promise.map(Object.keys(this.assignmentsByWorkerId), workerId => {
+      const key = this.getWorkerAssignmentsKey(workerId);
+      const connStrs = this.assignmentsByWorkerId[workerId];
+      return this.redisClient.saddAsync(key, connStrs)
+        .then(() => this.redisClient.expireAsync(key, this.masterDuration * 60));
+    });
+  }
+
+  recordAssignment(workerId, connStr) {
+    this.assignmentsByConnStr[connStr] = this.assignmentsByConnStr[connStr] || [];
+    this.assignmentsByConnStr[connStr].push(workerId);
+    this.assignmentsByWorkerId[workerId] = this.assignmentsByWorkerId[workerId] || [];
+    this.assignmentsByWorkerId[workerId].push(connStr);
+  }
+
+  getAvailableWorkerId() {
+    let availableWorkerId = this.workerId;
+    Object.keys(this.assignmentsByWorkerId).forEach(workerId => {
+      if (this.assignmentsByWorkerId[workerId].length < this.assignmentsByWorkerId[availableWorkerId]) {
+        availableWorkerId = workerId;
+      }
+    });
+    debug(`availableWorkerId: ${availableWorkerId}`);
+    return availableWorkerId;
+  }
+
+  populateWorkerAssignments() {
+    this.assignmentsByConnStr = {};
+    this.assignmentsByWorkerId = {};
+    return this.getWorkerIds()
+      .then(workerIds => {
+        return Promise.map(workerIds, workerId => {
+          return this.readWorkerAssignmentsFromRedis(workerId)
+            .then(connStrs => {
+              connStrs.forEach(connStr => this.recordAssignment(workerId, connStr));
+            });
+        });
+      });
```

Comment:
Too much nesting IMO; recommending splitting the readWorkerAssignmentsFromRedis block into a helper function.